gobblin-runtime/src/main/java/org/apache/gobblin/runtime/TaskStateCollectorService.java

```diff
@@ -235,6 +255,49 @@ public Void call() throws Exception {
     this.eventBus.post(new NewTaskCompletionEvent(ImmutableList.copyOf(taskStateQueue)));
   }
 
+  /**
+   * Uses the size of work units to determine a job's progress and reports the progress as a percentage via
+   * GobblinTrackingEvents
+   * @param taskState of job launched
+   */
+  private void reportJobProgress(TaskState taskState) {
+    String stringSize = taskState.getProp(ServiceConfigKeys.WORK_UNIT_BYTE_SIZE);
+    if (stringSize == null) {
+      LOGGER.warn("Expected to report job progress but work unit byte size property null");
+      return;
+    }
+
+    Long taskByteSize = Long.parseLong(stringSize);
+
+    // if progress reporting is enabled, value should be present
+    if (!this.jobState.contains(AbstractJobLauncher.TOTAL_BYTES_TO_COPY)) {
+      LOGGER.warn("Expected to report job progress but total bytes to copy property null");
+      return;
+    }
+    this.totalSizeToCopy = this.jobState.getPropAsLong(AbstractJobLauncher.TOTAL_BYTES_TO_COPY);
+
+    // avoid flooding Kafka message queue by sending GobblinTrackingEvents only when threshold is passed
+    this.bytesCopiedSoFar += taskByteSize;
```

Comment:
so the bytesCopiedSoFar will be updated once the first mapper completes? Is this true?
If yes, is it okay to show timeLeft = 0 before any mapper finished? Should we return "-1"  or something? Usually how long will it take for first mapper to complete?

Reply:
Yes it will be 0 until the first mapper completes but the percentage of progress will also be 0 - one will not be set without the other and we hope this is understandable enough to the user. In the usual case we expect the first mapper to complete in seconds or 1-2 min.

Reply:
right now we have this for jobEndTime as well I noticed in the jobStatistics (everything starts at 0)